multik-cuda/src/main/kotlin/org/jetbrains/kotlinx/multik/cuda/GpuArray.kt

```diff
@@ -0,0 +1,125 @@
+        checkResult(JCuda.cudaFree(deviceDataPtr))
+
+        isLoaded = false
+    }
+
+    companion object {
+        private const val CACHE_INITIAL_CAPACITY = 16
+        private const val CACHE_LOAD_FACTOR = 0.75f
+
+        private val cache = Collections.synchronizedMap(
+            LinkedHashMap<Any, GpuArray>(CACHE_INITIAL_CAPACITY, CACHE_LOAD_FACTOR, true)
+        )
+
+        private val deleteQueue = LinkedBlockingQueue<Any>()
+        private val cleaner = Cleaner.create()
+
+        fun assertAllLoaded(vararg arrays: GpuArray) {
+            if (!arrays.all { it.isLoaded })
+                throw OutOfMemoryError("Not all arrays have are loaded in the GPU memory")
+        }
+
+        fun fullCleanup() {
+            cache.forEach { it.value.free() }
+            cache.clear()
+
+            deleteQueue.clear()
+        }
+
+        fun <T : Number, D : Dimension> getOrAlloc(array: MultiArray<T, D>, setMemory: Boolean = true): GpuArray {
+            cleanup()
+            return cache.getOrPut(array.data.data) { allocMemory(array, setMemory) }
+        }
+
+        private fun getGpuMemInfo(): String {
+            val free = LongArray(1)
+            val total = LongArray(1)
+
+            checkResult(JCuda.cudaMemGetInfo(free, total))
+            return "free: ${byteSizeToString(free[0])}, total: ${byteSizeToString(total[0])} MB"
+        }
+
+        private fun <T : Number, D : Dimension> allocMemory(array: MultiArray<T, D>, setMemory: Boolean): GpuArray {
+            val elemSize = array.dtype.itemSize
+            val rawData = array.data.data
+
+            val deviceDataPtr = Pointer()
+            val hostDataPtr = array.dtype.getDataPointer(array.data)
+            val byteSize = elemSize.toLong() * array.size
+
+            logger.debug { "Allocating array on GPU. Data: ${rawData}, size: ${byteSizeToString(byteSize)}" }
+
+            while (true) {
+                val result = JCuda.cudaMalloc(deviceDataPtr, byteSize)
+
+                if (result == cudaErrorMemoryAllocation) {
+                    logger.trace { "Not enough GPU memory for allocation. Trying to free stale memory. Gpu Mem: {${getGpuMemInfo()}}" }
+
+                    val iterator = cache.iterator()
+                    if (iterator.hasNext()) {
+                        iterator.next().value.free()
+                        iterator.remove()
+                    } else {
+                        throw CudaException("Insufficient GPU memory for the array")
+                    }
+                } else
+                    break
+            }
+
+            if (setMemory)
+                checkResult(JCuda.cudaMemcpy(deviceDataPtr, hostDataPtr, byteSize, cudaMemcpyKind.cudaMemcpyHostToDevice))
+
+            cleaner.register(array.data) { deleteQueue.add(rawData) }
+
+            return GpuArray(rawData, hostDataPtr, deviceDataPtr, byteSize)
+        }
+
+        private fun cleanup() {
```

Comment:
I don't really understand the meaning of this method, could you explain it?

Reply:
Discussed in person